boot/bootutil/src/loader.c

```diff
@@ -1916,3 +1914,212 @@ split_go(int loader_slot, int split_slot, void **entry)
     free(sectors);
     return rc;
 }
+
+#else /* !MCUBOOT_XIP */
+
+/**
+ * Sorts the available images based on the version number and puts them in
+ * a list.
+ *
+ * @param state          Boot loader status information.
+ * @param boot_sequence  A pointer to an array, whose aim is to carry
+ *                       the boot order of candidate images.
+ * @param slot_cnt       The number of flash areas, which can contain firmware
+ *                       images.
+ *
+ * @return               The number of valid images.
+ */
+static uint32_t
+boot_get_boot_sequence(struct boot_loader_state *state,
+                       uint32_t *boot_sequence, uint32_t slot_cnt)
+{
+    struct boot_swap_state slot_state;
+    struct image_header *hdr = NULL;
+    struct image_slot_version image_versions[BOOT_NUM_SLOTS] = {{0}};
+    uint32_t image_cnt = 0;
+    uint32_t slot;
+    int32_t rc;
+    int32_t fa_id;
+
+    for (slot = 0; slot < slot_cnt; slot++) {
+        hdr = boot_img_hdr(state, slot);
+        fa_id = flash_area_id_from_image_slot(slot);
+        rc = boot_read_swap_state_by_id(fa_id, &slot_state);
+        if (rc != 0) {
+            BOOT_LOG_ERR("Error during reading image trailer from slot: %u",
+                         slot);
+            continue;
+        }
+
+        if (boot_is_header_valid(hdr, BOOT_IMG_AREA(state, slot))) {
+            if (slot_state.magic    == BOOT_MAGIC_GOOD ||
+                slot_state.image_ok == BOOT_FLAG_SET) {
+                /* Valid cases:
+                 *  - Test mode:      magic is OK in image trailer
+                 *  - Permanent mode: image_ok flag has previously set
+                 */
+                image_versions[slot].slot_number = slot;
+                image_versions[slot].version = boot_get_version_number(hdr);
+                image_cnt++;
+            }
+
+            BOOT_LOG_IMAGE_INFO(slot, hdr, &slot_state);
+
+        } else {
+            BOOT_LOG_INF("%s slot: No valid image", (slot == BOOT_PRIMARY_SLOT)
+                         ? "Primary" : "Secondary");
+        }
+    }
+
+    /* Sort the images based on version number */
+    qsort(image_versions, slot_cnt, sizeof(struct image_slot_version),
```

Comment:
Given that you're only comparing two things, pulling in qsort is kind of overkill. How about just an if statement and swap? Or even just build the boot sequence array as a result of the 'if'.

Reply:
> > First reaction is that I think we need a better name for this, since normal mode also executes in place, in fact that is why it has to move the images. I'll try to think of something.
> 
> `execute-in-place-on-secondary-slot`? 🤣

Original feature name was NO_SWAP in TF-M.

Reply:
I believe originally this solution was designed to be able to handle N number of slots.
However, if we limit the slot number to 2 (which is the current limit for the whole project)
and provide a solution for more slots later if it will be ever needed, we can easily replace qsort with an 'if' and the newly introduced 'image_slot_version' structure and boot_compare_version_numbers() function could be abandoned.
I think this is the weakest and most questionable part of the whole PR.

Reply:
@d3zd3z , @tamasban what do yo think?

Reply:
Have you checked the size of the qsort implementation being brought in? I would just casually guess it probably adds 3k or so to the image. I would say, even with 'n' images, we shouldn't be calling quicksort.

In fact, why not just scan it for the newest image, and then if that isn't valid, scan it again for the next biggest. It's technically n^2 but I can't imaging having n ever be very large. (I really don't see why it would be larger than 2).

Reply:
We can drop qsort if it has such a big flash requirements. I agree that probably always considering to have  two images is very reasonable.